server/src/main/java/org/opensearch/common/AdjustableSemaphore.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common;
+
+import java.util.concurrent.Semaphore;
+
+/**
+ * AdjustableSemaphore is Extended Semphore where we can change maxPermits.
+ */
+public class AdjustableSemaphore extends Semaphore {
+    private final Object maxPermitsMutex = new Object();
+    private int maxPermits;
+
+    public AdjustableSemaphore(int maxPermits) {
+        super(maxPermits);
+        this.maxPermits = maxPermits;
+    }
+
+    /**
+     * Update the maxPermits in semaphore
+     */
+    public void setMaxPermits(int permits) {
+        synchronized (maxPermitsMutex) {
+            final int diff = Math.subtractExact(permits, maxPermits);
+            if (diff > 0) {
+                // add permits
+                release(diff);
+            } else if (diff < 0) {
+                // remove permits
+                reducePermits(Math.negateExact(diff));
```

Comment:
Old PR Comment:[https://github.com/opensearch-project/OpenSearch/pull/484#discussion_r611457535]

what happens if those permits are already acquired? does it fail or wait for those permits to be released?

Reply:
Update call will pass even if all permits are acquired. New acquire will fail, it will pass once enough permits are not released and there are available permits as per new limit. 

For e.g. Limit is 8, acquired 8 permits, Changed limit to 5, release 3 permits, acquire of 1 permit wont be allowed as per new limit and there are already 5 permits are acquired.